src/scenechange/mod.rs

```diff
@@ -7,69 +7,202 @@
-    match self.last_frame {
-      Some((last_num, ref last_frame)) if last_num == frame_num - 1 => {
-        let len = curr_frame.planes[0].cfg.width * curr_frame.planes[0].cfg.height;
-        let delta_yuv = last_frame.iter().zip(curr_frame.iter())
-          .map(|(last, cur)| (
-            (i16::cast_from(cur.0) - i16::cast_from(last.0)).abs() as u64,
-            (i16::cast_from(cur.1) - i16::cast_from(last.1)).abs() as u64,
-            (i16::cast_from(cur.2) - i16::cast_from(last.2)).abs() as u64
-          )).fold((0, 0, 0), |(ht, st, vt), (h, s, v)| (ht + h, st + s, vt + v));
-        let delta_yuv = (
-          (delta_yuv.0 / len as u64) as u16,
-          (delta_yuv.1 / len as u64) as u16,
-          (delta_yuv.2 / len as u64) as u16
-        );
-        let delta_avg = ((delta_yuv.0 + delta_yuv.1 + delta_yuv.2) / 3) as u8;
-        delta_avg >= self.threshold
+  /// Runs dynamic scenechange detection on the frames in `frame_subset`.
+  /// Specifically, uses lookahead to exclude any frames that count as a short flash
+  /// of content, then checks the previous and current frame for differences
+  /// to determine if they qualify as a scenechange.
+  fn detect_scene_change<T: Pixel>(
+    &mut self,
+    frame_subset: &[Arc<Frame<T>>],
+    frameno: u64,
+    inter_cfg: &InterConfig,
+  ) -> bool {
+    if self.excluded_frames.contains(&frameno) {
+      // This frame has already been excluded as a candidate for keyframe selection.
+      return false;
+    }
+
+    // Lookahead to avoid coding short flashes as scenecuts
+    let lookahead_distance = cmp::min(
+      cmp::max(1, inter_cfg.pyramid_depth as usize) + 1,
+      frame_subset.len() - 1,
+    );
+    let mut excluded = false;
+
+    // Where A and B are scenes: AAAAAABBBAAAAAA
+    // If BBB is shorter than lookahead_distance, it is detected as a flash
+    // and not considered a scenecut.
+    if lookahead_distance > 1 {
+      for j in 1..=lookahead_distance {
+        if !self.has_scenecut(&frame_subset[0], &frame_subset[j]) {
+          // Any frame in between `0` and `j` cannot be a real scenecut.
+          for i in 0..=j {
+            let frameno = frameno + i as u64 - 1;
+            self.excluded_frames.insert(frameno);
+          }
+          excluded = true;
+        }
       }
-      _ => panic!("Incorrect frame order in SceneChangeDetector")
     }
+
+    // Where A-F are scenes: AAAAABBCCDDEEFFFFFF
+    // If each of BB ... EE are shorter than (lookahead_distance-p0), they are
+    // detected as flashes and not considered scenecuts.
+    // Instead, the first F frame becomes a scenecut.
+    // If the video ends before F, no frame becomes a scenecut.
+    for i in 1..=lookahead_distance {
+      if i < lookahead_distance
+        && self
+          .has_scenecut(&frame_subset[i], &frame_subset[lookahead_distance])
+      {
+        // If cur_p0 is the p0 of a scenecut, it cannot be the p1 of a scenecut.
```

Comment:
What do p0 and p1 mean here in rav1e's case?

Reply:
Good catch, will clarify.